### ISSUE TITLE
replace org.jmesa:jmesa-oc

### ISSUE THREAD

**Issue by chae-reliatec** [reliatec-gmbh/LibreClinica#74]
replace org.jmesa:jmesa-oc since this is available from the OpenClinica respository only.

**Comment by toskrip** [comment]
The idea is to use DataTables jQuery plug-in to provide client side rendering on all places where jmesa was used. On server site the REST API will be created compatible with DataTables server side processing requirements so that the tables perform well also on big datasets.

**Comment by toskrip** [comment]
should verify whether update to jmesa library (2.5.3) is possible

**Comment by toskrip** [comment]
one jmesa fork that could be also examined:

https://github.com/yujiaao/jmesa

**Comment by oba-reliatec** [comment]
RFC: Please look at @constructum's analysis and suggestion for jmesa replacement in the [wiki](https://github.com/reliatec-gmbh/LibreClinica/wiki/jmesa-replacement:-scope,-proposal-and-implementation-roadmap).

**Comment by toskrip** [comment]
@constructum would it be possible to commit the example implementation to lc-remove-jmesa branch? I understand the HtmlFlow part for server side rendering but I am not certain about how much of HTMX is actually needed and how would this actually fit the JSP stack? Is the intention there to have servlets returning partial HTML chunks or adding additional Spring REST controllers that return partial HTML? 

The biggest advantage of continuing with SSR is that it would be under our control and no dependency on specific datatable library, but this has a consequence that we are actually maintaining our own data table library rather then using existing components.

Having mix of jmesa and non-jmesa tables to gradually do the change sounds better to me. But I do not see much sense in having both implementation for one table in parallel at the same time and switching between with environment variable. This would mean that for each release we have to actually test both/ do smoke testing twice.

**Comment by constructum** [comment]
@toskrip Originally I planned to commit to a public branch after finishing the first table (`AuditUserActivity`), but--since you asked--I have now committed a *preliminary* version to the `lc-remove-jmesa` branch of my own fork, see here:
https://github.com/constructum/LibreClinica/commits/lc-remove-jmesa/

Note that it is not feature-complete yet, needs works on the styling, and the public interface of the 'lctable' package may change, but the basics are already in place. I published it to give you (and others) the opportunity to have a look and get an idea of how it works.

Concerning your other questions/remarks:
1. the role of HTMX is essentially to trigger a partial reload of the table, i.e. avoid a full page reload, when the table view changes (e.g. by navigation, filtering, etc.);
2. no additional servlets are needed: HTMX uses a special HTTP header `HX-Request` to distinguish between a regular request (to load a full page) and a HTMX request (for partial reload);
3. it is true that, going forward, we'll have to maintain our own data table library: therefore, I am trying to make it as simple and small as possible (I think it's going to be much simpler than jmesa);
4. I agree that it's best to merge the changes gradually;
5. the coexistence of both implementations is not meant to be in any release: it is only a temporary help for development (for example, I switch to jmesa when I want to check how the original table behaves); once the jmesa replacement is done and tested, jmesa is going to be completely removed (before release).

As a side remark: I have put the `lctable` package where the rest of the LC-web components are. However, since this new development fully originates from the LibreClinica community, what do you think about putting it in a `org.libreclinica.` namespace instead of `org.akaza.openclinica.` ?

**Comment by toskrip** [comment]
Tnx, much appreciated I will examine it hopefully sometime next week and come back if any issues. With the namespace change, although you are right that it is a new part of codebase but splitting the code like this does not feel natural. I mean namespace are for logical organization, we still use old namespaces in schemas and web services so unifying everything under one namespace would introduce breaking changes and having two namespaces next to each other logically splits the codebase for unnecessary reason in my opinion. I would not recommend that at this point.

**Comment by toskrip** [comment]
@constructum looks good! I see now how the HX-Requests are handled. It sounds promising.

**Comment by constructum** [comment]
@toskrip I am glad you approve of this approach, I'll continue along these lines then

**Comment by toskrip** [comment]
Small, thing. The language level compatibility in web module is now 11 (something that we will have to also adapt int devel documentation).

**Comment by constructum** [comment]
Yes, indeed. I have set it to 11 for the web module, but only in the `lc-remove-jmesa` branch (`web/pom.xml`).

This is necessary because I used features at this level to implement the new tables. It should not be a problem, as we already moved to Java 11.

But shouldn't the development documentation be adapted only after the `lc-develop` branch moves to level 11? Or are you proposing to move `lc-develop` to 11? (I do not completely understand what you intend to do about the  language level compatibility.)

**Comment by constructum** [comment]
User documentation for the LCTable library is now available in the project wiki, see
https://github.com/reliatec-gmbh/LibreClinica/wiki/LCTable-Library-User-Documentation

**Comment by toskrip** [comment]
> Yes, indeed. I have set it to 11 for the web module, but only in the `lc-remove-jmesa` branch (`web/pom.xml`).
> 
> This is necessary because I used features at this level to implement the new tables. It should not be a problem, as we already moved to Java 11.
> 
> But shouldn't the development documentation be adapted only after the `lc-develop` branch moves to level 11? Or are you proposing to move `lc-develop` to 11? (I do not completely understand what you intend to do about the language level compatibility.)

only disclosing it here for other devs that may want to try to debug the branch (my IDE refused to compile it without adapting the language level compatibility).

**Comment by toskrip** [comment]
So far looks pretty good. A couple of notes:

> LCTableColumnDef

we will need extra property for isFilterable (some tables have columns that are defined as not filterable)

> the timestamp validation

I really thing we are OK on client to do validation without leap year detection but rather full 4 digit year validity, at the end if non existing date is entered on server side it will be catch because conversion to date time will fail

> expecting UTC timezone for login attempts

This will not work. The login attempts timestamps are written to the database without timezone but they use local timezone configured on server.  Now when you expect that it is UTC, then first of all the render shows wrong time, the filter however matches the strings from database so this UTC not UTC inconsistency means that you can filter for something that you know exists as date time in database but the displayed results are converted to UTC....

> sort order UI indicator

probably not implemented yet, but I just put it here for completeness.

**Comment by constructum** [comment]
> > LCTableColumnDef
> 
> we will need extra property for isFilterable (some tables have columns that are defined as not filterable)

Columns without filter are already possible by setting, in column definitions, `filterDef` to `null` (or `NO_FILTER` for better readability):
```java
    public static final LCTableFilterDef NO_FILTER = null;
```
To address your remark, I added the following method:
```java
    public boolean isFilterable() { return this.filterDef != null; }
```

> 
> > the timestamp validation
> 
> I really thing we are OK on client to do validation without leap year detection but rather full 4 digit year validity, at the end if non existing date is entered on server side it will be catch because conversion to date time will fail

I agree that leap-year detection may be slightly over the top, but I still would like to catch dates that are obviously wrong on the client side to avoid sending unnecessary requests and receiving unnecessary error responses. We already agreed that a complex and unreadable regex is not the best way to do that. I'll think about a better way (not right now, but I'll put it in the to-do list).

> 
> > expecting UTC timezone for login attempts
> 
> This will not work. The login attempts timestamps are written to the database without timezone but they use local timezone configured on server. Now when you expect that it is UTC, then first of all the render shows wrong time, the filter however matches the strings from database so this UTC not UTC inconsistency means that you can filter for something that you know exists as date time in database but the displayed results are converted to UTC....

Good catch, thanks a lot! I did not notice this problem because I was testing the server in a container and the container uses UTC unless I instruct it otherwise. This is fixed now. I also changed a message in the Audit User Activity page to use 24h time format for consistency with the table, which also uses this format.

> 
> > sort order UI indicator
> 
> probably not implemented yet, but I just put it here for completeness.

They were already implemented, but a fix for another bug introduced this one (preventing the update of the sort order UI indicators). It's all explained in detail in the commit message.


The changes mentioned above are in PR https://github.com/reliatec-gmbh/LibreClinica/pull/459.

**Comment by constructum** [comment]
I am trying to put together a complete list of the jmesa tables to be migrated, to keep track of progress and to-do's. While doing that, I found out the following.

##### 1. `listDiscNotes` table

There is a table named `listDiscNotes`, constructed via `ListDiscNotesSubjectTableFactory` and used in `ListDiscNotesSubjectServlet`.

However, `http://localhost:8080/LibreClinica/ListDiscNotesSubjectServlet` returns `404 Page not found`.

It turns out that the corresponding `<servlet>` and `<servlet-mapping>` entries in `web/src/main/webapp/WEB-INF/web.xml` are commented out. Therefore, the servlet is unreachable, i.e. dead code.

Is this servlet abandoned or is it planned to activate it in the future?

##### 2. `sdv` table created by `renderSubjectsAggregateTable` method of `SDVUtil` class

There is a table named `sdv` (not the only one with this name), with no associated factory class, but created in the `renderSubjectsAggregateTable` method of `SDVUtil` class.

There are no callers for this method (according to IntelliJ "Find Usages"). Therefore, this one too seems to be dead code.

What is the situation concerning this table?

---

If one or both of the above are indeed abandoned, they could be deleted (and the respective tables would not need to be migrated and could be removed from the list).

**Comment by toskrip** [comment]
Yes, it is likely the case, When you detect possible unused view, skip the table implementation for it. You can also add it to this ticket #329 that is aggregating list to be examined and removed from sources.

**Comment by constructum** [comment]
Thanks for the clarification. The jmesa table migration list is now at
https://github.com/reliatec-gmbh/LibreClinica/wiki/jmesa-tables-migration:-progress-and-to-do.

I also entered `WEB-INF/jsp/managestudy/listSubjectDiscNote.jsp`(related to the first unused table mentioned in https://github.com/reliatec-gmbh/LibreClinica/issues/74#issuecomment-5008480457) as "unused view" in ticket https://github.com/reliatec-gmbh/LibreClinica/issues/329. 

With respect to the second table mentioned there, I am not sure whether there is a corresponding "view" in the sense of https://github.com/reliatec-gmbh/LibreClinica/issues/329. The code for this table could probably just be removed at the end of the jmesa migration along with all the other jmesa tables.

**Comment by oba-reliatec** [comment]
Just to give a bit of context: SDV stands for source data verification, a data control process used in clinical trials, and LibreClinica has this feature. You can find such a table at <http://localhost:8080/libreclinica/pages/viewAllSubjectSDVtmp>. I just noted the "tmp" at the end of the URL. If you cannot access it, log in as monitor.  So there should be some SDV table(s) left in the code, but maybe there are abandoned ones that can be deleted.

What do you mean by
> the second table mentioned there

\# 2 listDiscNotes? Again, some context. Discrepancy notes are notes in LibreClinica that can be attached to data entry fields in the forms. There is a flag next to each field to access/enter such notes, and the flag will change its colour dependent on the status. They are used in various contexts, e. g. in SDV: when a monitor detects an abnormality (e. g. body weight 888) he submits a discrepancy note to the researcher appearing in the SDV list, and next to the data entry field for weigh indicated by a certain colour of the flag.  
At <http://localhost:8080/libreclinica/ViewNotes> you can filter by subject, so maybe a page /ListDiscNotesSubjectServlet is not necessary any more.

**Comment by constructum** [comment]
I am sorry I wasn't clear: "the second table mentioned there" was meant to refer to the second table mentioned in [this comment](https://github.com/reliatec-gmbh/LibreClinica/issues/74#issuecomment-5008480457), i.e. the<br>"2. `sdv` table created by `renderSubjectsAggregateTable` method of `SDVUtil` class".

As there are no callers for the `renderSubjectsAggregateTable` method that constructs this table, the table is very likely unused in the LibreClinica web app. Therefore, the plan is: (1) do not migrate this table and then (2) after the migration of all tables is completed, if it is confirmed that this table is indeed unused, remove the `renderSubjectsAggregateTable` method.

**Comment by oba-reliatec** [comment]
Thank you for clarifying. Plan sounds good to me, once we arrive at step (2) look, whether it is used in LibreClinica-ws (just as a reminder).

**Comment by constructum** [comment]
Two questions about localisation.

I am trying to implement localisation properly, both in the LCTable library and in the specific tables that I am migrating. 

For the most part, I can reuse what I find in the factories of the jmesa classes, as the relevant properties are already defined (for example, in `ListStudySubjectTableFactory` all strings are properly localised and I can just reuse the relevant code for the new LCTable-based table).

However, sometimes I need to define new properties myself. This is the case for:
- the LCTable library itself
  (jmesa uses its own properties files `jmesaResourceBundle.properties`, which I obviously cannot use);
- some tables that are not properly localised in the existing code
  (for example, in `AuditUserLoginTableFactory` one can find lines such as
  `configureColumn(row.getColumn("userName"), "User Name", null, null)`:
  note the hard-coded column header title `"User Name"` ).

I have seen that there are properties files
1. `web/src/main/resources/org/akaza/openclinica/i18n/words.properties`
2. `core/src/main/resources/org/akaza/openclinica/i18n/words.properties`
3. `ws/src/main/resources/org/akaza/openclinica/i18n/words.properties`

I assume that the relevant one for me is the first one (`web/...`), can you confirm?

Second question: can I just add properties to `words.properties` as the need arises (after having checked that they do not already exist, of course)? Or are there specific procedures / guidelines to follow to do that?

**Comment by z10000** [comment]
@constructum — thanks for taking on the localisation. Here is what we found while working on
the Japanese translation (#447 / PR #448), in case it helps.

## Question 1: is `web/...` the relevant copy?

**Yes — the copy the web application actually reads at runtime is
`web/src/main/resources/...`.**

Inside each WAR, `WEB-INF/classes` (the module's own resources) shadows the copy inside
`WEB-INF/lib/LibreClinica-core.jar`. Checking the built WARs:

- `words.properties` resolves from `WEB-INF/classes` (i.e. from `web`); the copy in the core
  jar is unreachable there
- the `ws` WAR likewise reads its own copies, falling back to the core jar only for bundles it
  does not carry itself (`audit_events`)

This is also consistent with the fact that the translator-facing notes
(`#DO NOT CHANGE THE TEXT FOR …`) exist only in `web` and `ws`, and not once in `core`.

That said, there are three caveats that make the answer less clean than it looks.

1. **The three copies are not in sync and have drifted apart.** Between `core` and `web` there
   are **447 keys only in web, 40 keys only in core, and 81 keys whose values differ**
   (e.g. `add_new_event` is `Add New Event` in core but `Schedule New Event` in web).

2. **The 40 keys that exist only in `core` cannot be reached from the web application at all** —
   `footer.edition.1` and `last_study_deleted_1`, for instance, raise
   `MissingResourceException` in the English UI.

3. **Translations currently live in `core`, not in `web`.** So a key added to `web` is, by
   construction, invisible from where the translations are maintained.

In other words, the place where English keys are added (`web`) and the place where
translations live (`core`) are currently different, and every key added to `web` widens that
gap.

We have opened a separate issue proposing to remove the duplication itself
(#460): consolidate the English bundles into `core` and drop the `web`
copies, leaving `ws` frozen since it is legacy. Nothing changes visually for anyone. If that
is accepted, `core` becomes the place to add new keys.

**Suggestion for the meantime**: since having to move keys twice is extra work, it is probably
cleanest to **add new keys after the consolidation is decided**. If you would rather not wait,
adding them to `web` now is fine too — they can be moved together with the rest during the
consolidation, which is just a relocation.

## Question 2: may keys simply be added, and are there procedures?

**There is no documented procedure** — LibreClinica currently has no i18n documentation at
all. Whether keys may be added is presumably @toskrip's call, so all we can offer is the set
of conventions that *do* exist.

Those conventions live in the properties files themselves rather than in documentation. The
OpenClinica docs put it as "some strings should never be translated, and these are identified
with an appropriate comment in each of the properties files", and indeed:

| what | where | note |
|------|-------|------|
| Role names | `web/.../terms.properties`, **10 occurrences** | `#DO NOT CHANGE THE TEXT FOR DATA MANAGER` etc. (`Data_Entry_Person`, `Investigator`, `Monitor`, `Study_Coordinator`, `Study_Director`, `site_*`) |
| Licence text | `licensing.properties` | `# DO NOT CHANGE ANY OF THE TEXT BELOW` (LGPL, copyright, ReliaTec GmbH trademark notice) |
| Legal notices | `footer.license.*`, `footer.tooltip` in `words.properties` | not translated |

One more thing worth knowing when adding keys: **`format.properties` is not translation but
locale-dependent configuration**. Its own comments explain each property, but in short
`date_format` has to match `date_format_string`, and `date_regexp` is the regular expression
used to validate dates the user types in. It is handled differently from display labels.

Also note that everything under `src/main/resources` goes through **Maven resource
filtering**, so a `${...}` in a value is substituted at build time (existing example:
`version_number=${pom.version}` in `words.properties`).

## On localising LCTable

I read through the wiki documentation and the code on the branch. A few things stood out that
may be worth knowing before you implement, from the translation side.

### 1. The current structure cannot reflect the request's locale

This is probably the most significant one. Replacing `columnDisplayName` with a lookup into a
resource bundle will not be enough as things stand:

- `COLUMNS` in `AuditUserLoginTable` is `static final`
- the table instance is created **once**, in `AuditUserActivityServlet.init()`

so the column headers are evaluated a single time at servlet initialisation, and every user
gets the same strings regardless of their language. The old implementation
(`AbstractTableFactory`) took `LocaleResolver.getLocale(request)` inside
`createTable(request, response)` and called `configureColumns(tableFacade, locale)` per
request, so this is a structural difference.

Possible directions:

- **A**: hold a **bundle key** in `columnDisplayName` instead of a display string, and resolve
  it with the locale at `render()` time (a change in the library)
- **B**: build and cache `COLUMNS` per locale
- **C**: stop creating the instance in `init()` and create it per request (same behaviour as
  the old implementation)

A looks the most principled to me, but that is your call.

### 2. A caveat if you add a new bundle

If you create a new bundle for LCTable (say `lctable.properties`) and register it in
`ResourceBundleProvider` and `LocaleResolver.isQualifiedLocale()`, please add the translated
variants (`_ja` etc.) in the same change.

`isQualifiedLocale()` requires *all* listed bundles to resolve for a language, so a new bundle
without a `_ja` file would make **the entire Japanese UI fall back to English** — i.e. it
would silently disable the translation added in #448.

Adding keys to the existing `words` bundle avoids this entirely.

### 3. Much of it is already covered by existing keys

`words.properties` already contains `actions`, `apply_filter`, `hide`, `show_more`,
`select_an_event` and similar, all of which are already translated. The established pattern in
the existing code is:

```java
ResourceBundle reswords = ResourceBundleProvider.getWordsBundle(locale);
… reswords.getString("show_more") …
```

(`ListStudySubjectTableToolbar` lines 33/90/137/165, `AuditUserLoginTableFactory` lines 91/62)

What genuinely seems to need new keys is the handful of strings the LCTable library itself
carries:

- `"No results."` and `format("Results %d-%d of %d.", …)` in `LCTable.java`
- `"Rows: "` in `LCTable.java`
- `"Clear Filter"` in `LCTableFilterDef.java`
- `TIMESTAMP_FILTER_MESSAGE` in `LCTableUtil.java`

For reference, jmesa's built-in bundle has 30 keys in total, and the corresponding ones are
`html.statusbar.noResultsFound` and `html.statusbar.resultsFound=Results {1} - {2} of {0}.` —
possibly useful as a naming template. Note that the `format(...)` call would need to become a
`MessageFormat`-style `{0}` pattern when localised.

### 4. Smaller points

- The old `AuditUserLoginTableFactory` localised the header via
  `resword.getString("actions")` (line 62); the new implementation uses the literal
  `"Actions"`, so that one translation is currently lost.
- `LCTableFilterDef.Select.parseParam` matches the URL parameter against `label(val)`, i.e.
  the **display label** (line 142). It does not appear to be called from within the library at
  the moment, but once display labels are translated and this is used, the matching would
  break — comparing against `urlParam` would be safer. Since `enumCol` already lets you pass
  `renderer` and `urlParamConverter` separately, translating only the display text while
  keeping `name` in the URL works fine.
- `TIMESTAMP_FILTER_MESSAGE` embeds `yyyy-MM-dd`, whereas LibreClinica's `format.properties`
  has `date_format = DD-MMM-YYYY`. Spelling the filter format out in the message may end up
  duplicating that setting.

---

Hope some of this is useful.

**Comment by toskrip** [comment]
@constructum we will try to establish all i18n properties in core. Please add necessary terms in words.properties and also words_ja.properties using english that can later be translated.

**Comment by z10000** [comment]
@toskrip @constructum — we are glad to see the plan to establish all i18n properties in `core`.
One note on `words_ja.properties`, from the side that maintains the Japanese translation.

## On the intent

We read the instruction to add the terms to `words_ja.properties` as well as a way of keeping the
key sets aligned so that nothing gets missed later. It may also come from something we wrote
ourselves in #447:

> The Japanese files carry matching translations under the same keys so the key sets stay exactly
> aligned.

That sentence described a *result*: every existing key had been translated, so the key sets
happened to match. What we aligned back then were things like the corrupted leftover keys on the
English side (`notes` has `rrent_active_study_is` next to the real `your_current_active_study_is`),
where we placed a matching translation to keep the sets 1:1. **It does not carry over to adding new
English keys.**

## Technically the addition is not needed

If a key is absent from `words_ja`, the Japanese UI shows the English value through the
resource-bundle parent chain. That is by design.

**278 keys are in exactly that state today** — `403_error_msg_header`, `Assigned_to`,
`Audit_History` and so on — and we have confirmed in a browser that they render in English under a
Japanese locale. In other words, whether or not the English text is written into `words_ja`, the
rendered result is the same.

## Writing it in makes missing translations harder to find

- **Untranslated keys stop being detectable.** Diffing the key set of the English file against the
  `_ja` file relies on the key being *absent* to signal "not translated yet". If it is present with
  an English value, it does not show up in the diff and cannot be found mechanically.
- **It becomes indistinguishable from text that is meant to stay English.** `words_ja` currently
  holds 51 non-Japanese values, including code values such as `ASKU`, `PHI`, `CRF`, `OID`, and the
  legal notices `footer.license.*` which must not be changed. Once untranslated English is mixed in,
  telling the two apart requires going through them one by one.
- **It does not follow the English side.** With the key absent, the current English text is always
  shown. With English hard-coded into `words_ja`, a later change to the English wording leaves the
  old text in the Japanese UI. There is a real example of this: `version_number` sits in `words_ja`
  as `1.4.0rc1` while the English side has `${pom.version}`. That is a defect in our own PR #448,
  and we will drop the key in the consolidation PR (#460).

## Suggestion

**Please add the keys to `words.properties` only; `words_ja.properties` can be left untouched.**

Translations do not have to follow in lockstep. `LocaleResolver.isQualifiedLocale()` only checks
that the nine files exist for the language, so individual untranslated keys do not affect the
Japanese locale — those keys simply render in English.

We will document this convention in the README that goes with the consolidation PR.

**Comment by constructum** [comment]
Thanks @z10000 for the very thorough and insightful investigation, which is extremely helpful to me, and @toskrip for final directions.

I am going to wait until things are settled (resource bundle consolidation into core) before implementing localisation in the tables.

My current understanding is that, when implementing localisation (after the consolidation), I should add any new English properties that may be needed to `core/src/main/resources/org/akaza/openclinica/i18n/words.properties` (and nowhere else).

The difference in key set between `words.properties` and `words_ja.properties` will provide an indication that there is something that needs to be translated into Japanese. Nevertheless, I'd also advertise the addition of new tables-related properties via comment in https://github.com/reliatec-gmbh/LibreClinica/issues/74 (this issue), so that you can get immediate notification that there is something that may need translation.

Could you please confirm that I should proceed this way (or correct me / propose alternative way)?

**Comment by toskrip** [comment]
works for me, @constructum go for `words.properties`, when the fallback to english is working that is sufficient.

I had a quick look on the subjects matrix table, also looks pretty good. I know that you are trying to stay close to original but I would reconsider the implementation of show more columns. The old one was very restrictive allowing basically just two (basic vs extended) view. When we are already implementing new table rendering, we should consider changing the design and allow dynamic show hide columns selection on top of the standard default view.

**Comment by toskrip** [comment]
@constructum one issue I found, I think that the enforcement in web to produce java 11 bytecode while compiling with jdk 11 is no longer compatible with jersey 1.19.3. Restful URLs, part of web services are using jersey servlet and this seems to not work anymore. This is verifyable in ViewStudySubject any Subject Casebooks (html, json, odm xml) while navigate to the restful URL will lead to exception.

**Comment by constructum** [comment]
Thanks, @toskrip, for the very helpful feedback.

1. Concerning the localisation, I am going to wait for the consolidation into `core` (https://github.com/reliatec-gmbh/LibreClinica/issues/460, https://github.com/reliatec-gmbh/LibreClinica/pull/465) and then add what is needed to the `words.properties` in `core`.

2. Concerning the "hidden columns", I am open to different solutions. I only stayed so close to the original because, not knowing the preferences of the user community, this allows to me to progress more quickly towards a fully functional jmesa replacement. I'd propose to rediscuss this issue in September after the summer break (if we change something, it should be good, so I wouldn't like to rush it).

3. I had noticed the problem with the "Subject Casebook" and started investigating it more closely to write an issue. In the process, I noticed that it occurs only in the `lc-remove-jmesa` branch and I was quite puzzled, because that code should really be independent from the tables. Great that you found the cause! Do you already have an idea what can be done about it? Going back to 9 is not really a good option because the code for the tables relies on language features of Java 11. Of course, everything can be done, but that time-consuming and not productive. I'd really like to avoid it, if possible.

**Comment by toskrip** [comment]
Without modernization of jersey stack we can do something like this to tell it not to use ASM to detect all service classes in namespace (which otherwise fails on bytecode compatibility):

```xml
<servlet>
        <servlet-name>OpenClinicaJersey</servlet-name>
        <servlet-class>com.sun.jersey.spi.spring.container.servlet.SpringServlet</servlet-class>
<!--        <init-param>-->
<!--            <param-name>com.sun.jersey.config.property.packages</param-name>-->
<!--            <param-value>org.akaza.openclinica.web.restful</param-value>-->
<!--        </init-param>-->
        <init-param>
            <param-name>com.sun.jersey.config.property.classnames</param-name>
            <param-value>
                org.akaza.openclinica.web.restful.ClinicalDataCollectorResource
                org.akaza.openclinica.web.restful.JSONClinicalDataPostProcessor
                org.akaza.openclinica.web.restful.MetadataCollectorResource
                org.akaza.openclinica.web.restful.ODMClinicaDataResource
                org.akaza.openclinica.web.restful.ODMMetadataRestResource
            </param-value>
        </init-param>
        <init-param>
            <param-name>com.sun.jersey.spi.container.ContainerRequestFilters</param-name>
            <param-value>org.akaza.openclinica.web.filter.rest.RestODMFilter</param-value>
        </init-param>
        <load-on-startup>1</load-on-startup>
    </servlet>
````

the same (listing all the endpoint classes) would need to be done for OpenClinicaJersey2 which are services used in ePRO.

**Comment by constructum** [comment]
Thanks a lot @toskrip, it worked perfectly. I just submitted a pull request with the fix.

**Comment by constructum** [comment]
I intend to include `data-test-*` attributes in the HTML of the new tables to facilitate testing (these attributes make it easier to locate, within a table, the HTML elements that are needed by a test, while at the same time not interfering with anything else).

However, it is probably better not to include these attributes when the application runs in the production environment.

Is there a standard way in LibreClinica to check whether the application is running in the development or in the production environment? Or at least a good way? (I can think of a few ways, but they may not be the "right" ones, so I thought it might better to ask.)

**Comment by toskrip** [comment]
Personally I would think that it would be beneficial to have them (in consisten manner) present also in production so I would not distinguish between development and production for this purpose. The thing is that in production environment people wants do do UI automation to execute productional system validation. You can have a word with @oba-reliatec about this.

**Comment by constructum** [comment]
Thanks @toskrip, I see the point, it totally makes sense.

Also, as I am developing the test attributes design for the tables, I am noticing that you do not really need a lot of them, if the design is done properly, and they convey useful information. Conclusion: there is really no good reason to suppress them in production. Problem solved.
